### PR TITLE
Generate: basic token streaming

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Adds token streaming to `.generate()` 🎉 

### Why now? 
I want to showcase and communicate how much faster assisted generation can be... and for that, I need token streaming :D Non-image/video results have a much lower impact.

### What's being added 
This PR adds a `streamer` input to generate. If it is non-`None`, generate will call `streamer.put(new_tokens)` as they are being generated. `streamer` can, therefore, be a wide array of things. This PR adds the simplest case: print tokens as they are generated.

At first, I thought of adding a simpler `stream=True` option. However, the tokenizer would have to be passed into `.generate()`, which we have been avoiding, and it wouldn't be nearly as flexible. I've made the call to make streaming+`.generate()` flexible, and to keep it simple at a `pipeline` level.

### If this PR gets accepted
The plan is to:
1. Communicate this feature on Twitter (w/Colab examples)
2. Add to pipelines, maybe with a simpler `stream=True` flag to start
3. Add Gradio examples (and, if needed, a specific streamer class)
4. Add the beam search case to the streamer classes (beam search is much trickier -- we should only print tokens when all candidate beams agree, which means logic needs to be added)

### How does it look
Here's an example. Note that it is running on CPU, so we can actually see the streaming effect (3090 is too fast 😅 ). On GPU it also streams, but much faster 🔥 

https://user-images.githubusercontent.com/12240844/228595317-0f234e95-bd39-43a5-83e5-ef620da08eb0.mov

### CODE REVIEW

File: src/transformers/__init__.py [hunk 1]
```diff
@@ -934,6 +934,7 @@
             "StoppingCriteria",
             "StoppingCriteriaList",
             "TemperatureLogitsWarper",
+            "TextStreamer",
```
Comment (sgugger) [inline src/transformers/__init__.py:937]: Doesn't need to be gated by PyTorch.

**Review by sgugger — COMMENTED** [review]
I am not a fan of this API at all. Why is there a need for the `TextStreamer` to spawn a new process? The `put` method could directly call the print statement.

**Comment by gante** [comment]
@sgugger revised with the simpler implementation (no context manager nor multiprocessing) 🤗

**Review by sgugger — APPROVED** [review]
Better this way, thanks!

**Comment by oobabooga** [comment]
Just a FYI: I have been doing this using `transformers.StoppingCriteria` to create a callback:

```python
class Stream(transformers.StoppingCriteria):
    def __init__(self, callback_func=None):
        self.callback_func = callback_func

    def __call__(self, input_ids, scores) -> bool:
        if self.callback_func is not None:
            self.callback_func(input_ids[0])
        return False
```

The callback is then used to create an iterator with the Iteratorize class here: https://github.com/oobabooga/text-generation-webui/blob/main/modules/callbacks.py#L42

Usage becomes:

```python
def generate_with_callback(callback=None, **kwargs):
    kwargs['stopping_criteria'].append(Stream(callback_func=callback))
    with torch.no_grad():
        shared.model.generate(**kwargs)

def generate_with_streaming(**kwargs):
    return Iteratorize(generate_with_callback, kwargs, callback=None)

with generate_with_streaming(**generate_params) as generator:
    for output in generator:
```

**Comment by gante** [comment]
@oobabooga 🧠 That's a smart (and [unexpected](https://www.hyrumslaw.com/)!) use of the stopping criteria.

I'm going to work on a standardized Gradio solution today, and a Queue+iterator was indeed my plan. If you don't mind, I will take inspiration in your code 💛  

A question regarding your implementation -- you use a separate thread in the `Iteratorize`, not a separate process. Any reason for in picking a thread over a process? (Without running the code, I'd argue in favor of a separate thread for GIL purposes)

**Pull request merged**